tikv/kv_test.go

```diff
@@ -105,22 +124,127 @@ func (c *storeSafeTsMockClient) CloseAddr(addr string) error {
+	s.Require().Equal(uint64(100), ts)
+}
+
+func (s *testKVSuite) TestMinSafeTsFromPD() {
+	mockClient := storeSafeTsMockClient{Client: s.store.GetTiKVClient(), testSuite: s}
+	s.store.SetTiKVClient(&mockClient)
+	s.setGetMinResolvedTSByStoresIDs(func(ctx context.Context, ids []uint64) (uint64, map[uint64]uint64, error) {
+		return 90, nil, nil
+	})
+	until(10, time.Second, func() bool {
+		ts := s.store.GetMinSafeTS(oracle.GlobalTxnScope)
+		s.Require().False(math.MaxUint64 == ts)
+		return ts == 90
+	})
+	s.Require().Equal(atomic.LoadInt32(&mockClient.requestCount), int32(0))
+	s.Require().Equal(uint64(90), s.store.GetMinSafeTS(oracle.GlobalTxnScope))
+}
+
+func (s *testKVSuite) TestMinSafeTsFromPDByStores() {
+	mockClient := storeSafeTsMockClient{Client: s.store.GetTiKVClient(), testSuite: s}
+	s.store.SetTiKVClient(&mockClient)
+	s.setGetMinResolvedTSByStoresIDs(func(ctx context.Context, ids []uint64) (uint64, map[uint64]uint64, error) {
+		m := make(map[uint64]uint64)
+		for _, id := range ids {
+			m[id] = uint64(100) + id
+		}
+		return math.MaxUint64, m, nil
+	})
+	until(10, time.Second, func() bool {
+		ts := s.store.GetMinSafeTS(oracle.GlobalTxnScope)
+		s.Require().False(math.MaxUint64 == ts)
+		return ts == uint64(100)+s.tikvStoreID
+	})
+	s.Require().Equal(atomic.LoadInt32(&mockClient.requestCount), int32(0))
+	s.Require().Equal(uint64(100)+s.tikvStoreID, s.store.GetMinSafeTS(oracle.GlobalTxnScope))
+}
+
+func (s *testKVSuite) TestMinSafeTsFromMixed1() {
+	mockClient := storeSafeTsMockClient{Client: s.store.GetTiKVClient(), testSuite: s}
+	s.store.SetTiKVClient(&mockClient)
+	s.setGetMinResolvedTSByStoresIDs(func(ctx context.Context, ids []uint64) (uint64, map[uint64]uint64, error) {
+		m := make(map[uint64]uint64)
+		for _, id := range ids {
+			if id == s.tiflashStoreID {
+				m[id] = 0
+			} else {
+				m[id] = uint64(10)
+			}
+		}
+		return math.MaxUint64, m, nil
+	})
+	until(10, time.Second, func() bool {
+		ts := s.store.GetMinSafeTS("z1")
+		s.Require().False(math.MaxUint64 == ts)
+		return ts == uint64(10)
+	})
+	s.Require().GreaterOrEqual(atomic.LoadInt32(&mockClient.requestCount), int32(1))
+	s.Require().Equal(uint64(10), s.store.GetMinSafeTS(oracle.GlobalTxnScope))
+	s.Require().Equal(uint64(10), s.store.GetMinSafeTS("z1"))
+	s.Require().Equal(uint64(80), s.store.GetMinSafeTS("z2"))
+}
+
+func (s *testKVSuite) TestMinSafeTsFromMixed2() {
+	mockClient := storeSafeTsMockClient{Client: s.store.GetTiKVClient(), testSuite: s}
+	s.store.SetTiKVClient(&mockClient)
+	s.setGetMinResolvedTSByStoresIDs(func(ctx context.Context, ids []uint64) (uint64, map[uint64]uint64, error) {
+		m := make(map[uint64]uint64)
+		for _, id := range ids {
+			if id == s.tiflashStoreID {
+				m[id] = uint64(10)
+			} else {
+				m[id] = math.MaxUint64
+			}
+		}
+		return math.MaxUint64, m, nil
+	})
+	until(10, time.Second, func() bool {
+		ts := s.store.GetMinSafeTS("z1")
+		s.Require().False(math.MaxUint64 == ts)
+		return ts == uint64(10)
```

Comment:
This `until` call will timeout and the predicate never returns true.

Reply:
Should we check the return value of `until`?

Reply:
Sorry, it should GetMinSafeTS from z2.